src/erp12/ga_clj/toolbox.cljc

```diff
@@ -42,6 +77,30 @@
     (assoc gn idx2 (gn idx1) idx1 (gn idx2))))
 
 
+(defn uniform-addition
+  [genome addition-rate genetic-source]
+  (apply concat
+         (map #(if (< (rand) addition-rate)
+                 (if (< (rand) 0.5)
+                   (list % (rand-nth genetic-source))
+                   (list (rand-nth genetic-source) %))
+                 (list %))
+              genome)))
+
+(defn uniform-deletion
+  [genome deletion-rate]
+  (random-sample (- 1.0 deletion-rate) genome))
+
+(defn umad
+  "Performs uniform mutation by addition and deletion.
+   First pass adds a new gene before or after each gene in the genome.
+   Second pass deletes genes with some probability."
+  [genome addition-rate deletion-rate genetic-source]
```

Comment:
I hadn't thought about this until now, but do you think we should standardize the interface of each kind of algorithm? I'm noticing that all mutation functions take at least a genome as argument, but possibly other values as well. This makes it somewhat difficult to swap one mutation for another.

Suppose we standardize to:
- Selection is `population -> individual`
- Mutation is `individual -> genome`
- Recombination is `(coll of individual) -> genome`

We already conform to this with selection by passing config to `make-*` and getting back the selection method function in return. Would it be valuable to follow this pattern for mutation and recombination as well?

Reply:
This seems like a good idea. The alternative is to pass around a map with all of the parameters for everything, like we do in Clojush, which seems like a bad idea. Might as well use closures everywhere as long as we're using Clojure!

Should we generalize further and just say that all genetic operators take a collection of individuals, it's just that mutation will be given a collection containing 1 individual, crossover will contain 2, and weird recombination X will take N individuals?

Reply:
I thought about this too and I am on the fence. On one hand, I like that it would be more general. On the other hand, it seems like it might result in a lot of wrapping and unwrapping of individuals. Consider the `:breed` function in the example:

```clojure
(fn [population]
  (->> (repeatedly 2 #(lexicase population))
       (map :genome)
       tb/uniform-crossover
       tb/swap-2-genes))
```
would become
```clojure
(fn [population]
  (->> (repeatedly 2 #(lexicase population))
       (map :genome)
       tb/uniform-crossover
       vector ;; <--  Is this confusing to readers?
       tb/swap-2-genes))
```
Another example: a UMAD only breed function would be
```clojure
(fn [population] (umad [(lexicase population)])
;; OR
(fn [population] (umad (lexicase population))

```
I could be convinced either way.

Reply:
Yup, I see that. Though, uniform crossover currently takes a collection of individuals, not exactly 2 genomes. I suppose you could say it could take a collection of more than 2 individuals and still work?

Reply:
It can operate on more than 2 genomes. I think I lifted that from one of our other projects, but I forget which.
I left an `@todo` to add tests to those functions, when I get around to that I will be sure to tests the case with >2 parents.